contk/metric/metric.py

```diff
@@ -202,6 +205,131 @@ def close(self):
+				Contains start token (eg: ``<go>``) and end token (eg: ``<eos>``).
+				Size: `[batch_size, max_sentence_length]`
+			data[gen_key] (list or :class:`numpy.array` of `int`): Setences generated by model.
+				Contains end token (eg: ``<eos>``), but without start token (eg: ``<go>``).
+				Size: `[batch_size, gen_sentence_length]`.
+		'''
+		gen = data[self.gen_key]
+
+		for gen_sen in gen:
+			self.hyps.append(self.dataloader.trim_index(gen_sen))
+
+	def run_f(self, ele):
+		reference, fn = ele
+		BLEUscore_f = sentence_bleu(reference, fn, smoothing_function=SmoothingFunction().method1)
+		return BLEUscore_f
+
+	def close(self):
+		st = time.time()
+		'''Return a dict which contains:
+
+			* **bleu**: bleu value.
+		'''
+		#################################################
+		SAMPLE = 1000
+		random.shuffle(self.hyps)
+		ref2 = self.hyps[:SAMPLE]
+		pool = Pool(multiprocessing.cpu_count())
+		bleu_irl = pool.map(self.run_f, [(ref2[:i]+ref2[i+1:], ref2[i]) for i in range(SAMPLE)])
+		pool.close()
+		pool.join()
+		print(time.time()-st)
+		try:
+			return {"bleu":  1.0 * sum(bleu_irl) / len(bleu_irl)}
+		except ZeroDivisionError as _:
+			raise ZeroDivisionError("Bleu smoothing divided by zero. This is a known bug of corpus_bleu, \
+				usually caused when there is only one sample and the sample length is 1.")
+
+class FwBwBleuCorpusMetric(MetricBase):
+	'''Metric for calcualting BLEU.
+
+	Arguments:
+		reference_key (str): Reference sentences are passed to :func:.forward by ``data[reference_key]``.
+			Default: ``resp``.
+		gen_key (str): Sentences generated by model are passed to :func:.forward by
+			``data[gen_prob_key]``. Default: ``gen``.
+	'''
+	def __init__(self, dataloader, reference_key="resp_gan", gen_key="gen_gan"):
+		super().__init__()
+		self.dataloader = dataloader
+		self.reference_key = reference_key
+		self.gen_key = gen_key
+		self.refs = []
+		self.hyps = []
+
+	def forward(self, data):
+		'''Processing a batch of data.
+
+		Arguments:
+			data (dict): A dict at least contains the following keys.
+			data[reference_key] (list or :class:`numpy.array` of `int`): Reference sentences.
+				Contains start token (eg: ``<go>``) and end token (eg: ``<eos>``).
+				Size: `[batch_size, max_sentence_length]`
+			data[gen_key] (list or :class:`numpy.array` of `int`): Setences generated by model.
+				Contains end token (eg: ``<eos>``), but without start token (eg: ``<go>``).
+				Size: `[batch_size, gen_sentence_length]`.
+		'''
+		gen = data[self.gen_key]
+		resp = data[self.reference_key]
+
+		for gen_sen, resp_sen in zip(gen, resp):
+			self.hyps.append(self.dataloader.trim_index(gen_sen))
+			self.refs.append(self.dataloader.trim_index(resp_sen[1:]))
+
+	def run_f(self, ele):
+		reference, fn = ele
+		BLEUscore_f = sentence_bleu(reference, fn, smoothing_function=SmoothingFunction().method1)
+		return BLEUscore_f
+
+	def close(self):
+		'''Return a dict which contains:
+
+			* **bleu**: bleu value.
+		'''
+		SAMPLE = 1000
```

Comment:
This hyper-parameter should be the arguments of `__init__` and should be checked to avoid overflow.